File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/RoutingIdsAndExplicitDetails.cpp
// RoutingIdsAndExplicitDetails.cpp 
#include "RoutingIdsAndExplicitDetails.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

RoutingIdsAndExplicitDetails::RoutingIdsAndExplicitDetails(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //routingIdsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* routingIdsNode = xmlNode->FirstChildElement("routingIds");

   if(routingIdsNode){routingIdsIsNull_ = false;}
   else{routingIdsIsNull_ = true;}

   if(routingIdsNode)
   {
      for(routingIdsNode; routingIdsNode; routingIdsNode = routingIdsNode->NextSiblingElement("routingIds")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingIdsNode , address : " << routingIdsNode << std::endl;
          #endif
          if(routingIdsNode->Attribute("href") || routingIdsNode->Attribute("id"))
          {
              if(routingIdsNode->Attribute("id"))
              {
                  routingIdsIDRef_ = routingIdsNode->Attribute("id");
                  routingIds_.push_back(boost::shared_ptr<RoutingIds>(new RoutingIds(routingIdsNode)));
                  routingIds_.back()->setID(routingIdsIDRef_);
                  IDManager::instance().setID(routingIdsIDRef_, routingIds_.back());
              }
              else if(routingIdsNode->Attribute("href")) { routingIdsIDRef_ = routingIdsNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { routingIds_.push_back(boost::shared_ptr<RoutingIds>(new RoutingIds(routingIdsNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingIdsNode , address : " << routingIdsNode << std::endl;
       #endif
   }

   //routingNameNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* routingNameNode = xmlNode->FirstChildElement("routingName");

   if(routingNameNode){routingNameIsNull_ = false;}
   else{routingNameIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingNameNode , address : " << routingNameNode << std::endl;
   #endif
   if(routingNameNode)
   {
      if(routingNameNode->Attribute("href") || routingNameNode->Attribute("id"))
      {
          if(routingNameNode->Attribute("id"))
          {
             routingNameIDRef_ = routingNameNode->Attribute("id");
             routingName_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(routingNameNode));
             routingName_->setID(routingNameIDRef_);
             IDManager::instance().setID(routingNameIDRef_,routingName_);
          }
          else if(routingNameNode->Attribute("href")) { routingNameIDRef_ = routingNameNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { routingName_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(routingNameNode));}
   }

   //routingAddressNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* routingAddressNode = xmlNode->FirstChildElement("routingAddress");

   if(routingAddressNode){routingAddressIsNull_ = false;}
   else{routingAddressIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingAddressNode , address : " << routingAddressNode << std::endl;
   #endif
   if(routingAddressNode)
   {
      if(routingAddressNode->Attribute("href") || routingAddressNode->Attribute("id"))
      {
          if(routingAddressNode->Attribute("id"))
          {
             routingAddressIDRef_ = routingAddressNode->Attribute("id");
             routingAddress_ = boost::shared_ptr<Address>(new Address(routingAddressNode));
             routingAddress_->setID(routingAddressIDRef_);
             IDManager::instance().setID(routingAddressIDRef_,routingAddress_);
          }
          else if(routingAddressNode->Attribute("href")) { routingAddressIDRef_ = routingAddressNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { routingAddress_ = boost::shared_ptr<Address>(new Address(routingAddressNode));}
   }

   //routingAccountNumberNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* routingAccountNumberNode = xmlNode->FirstChildElement("routingAccountNumber");

   if(routingAccountNumberNode){routingAccountNumberIsNull_ = false;}
   else{routingAccountNumberIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingAccountNumberNode , address : " << routingAccountNumberNode << std::endl;
   #endif
   if(routingAccountNumberNode)
   {
      if(routingAccountNumberNode->Attribute("href") || routingAccountNumberNode->Attribute("id"))
      {
          if(routingAccountNumberNode->Attribute("id"))
          {
             routingAccountNumberIDRef_ = routingAccountNumberNode->Attribute("id");
             routingAccountNumber_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(routingAccountNumberNode));
             routingAccountNumber_->setID(routingAccountNumberIDRef_);
             IDManager::instance().setID(routingAccountNumberIDRef_,routingAccountNumber_);
          }
          else if(routingAccountNumberNode->Attribute("href")) { routingAccountNumberIDRef_ = routingAccountNumberNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { routingAccountNumber_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(routingAccountNumberNode));}
   }

   //routingReferenceTextNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* routingReferenceTextNode = xmlNode->FirstChildElement("routingReferenceText");

   if(routingReferenceTextNode){routingReferenceTextIsNull_ = false;}
   else{routingReferenceTextIsNull_ = true;}

   if(routingReferenceTextNode)
   {
      for(routingReferenceTextNode; routingReferenceTextNode; routingReferenceTextNode = routingReferenceTextNode->NextSiblingElement("routingReferenceText")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingReferenceTextNode , address : " << routingReferenceTextNode << std::endl;
          #endif
          if(routingReferenceTextNode->Attribute("href") || routingReferenceTextNode->Attribute("id"))
          {
              if(routingReferenceTextNode->Attribute("id"))
              {
                  routingReferenceTextIDRef_ = routingReferenceTextNode->Attribute("id");
                  routingReferenceText_.push_back(boost::shared_ptr<XsdTypeString>(new XsdTypeString(routingReferenceTextNode)));
                  routingReferenceText_.back()->setID(routingReferenceTextIDRef_);
                  IDManager::instance().setID(routingReferenceTextIDRef_, routingReferenceText_.back());
              }
              else if(routingReferenceTextNode->Attribute("href")) { routingReferenceTextIDRef_ = routingReferenceTextNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { routingReferenceText_.push_back(boost::shared_ptr<XsdTypeString>(new XsdTypeString(routingReferenceTextNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- routingReferenceTextNode , address : " << routingReferenceTextNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
std::vector<boost::shared_ptr<RoutingIds>> RoutingIdsAndExplicitDetails::getRoutingIds()
{
   if(!this->routingIdsIsNull_){
        if(routingIdsIDRef_ != ""){
             return this->routingIds_;
        }else{
             return this->routingIds_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<RoutingIds>>();
   }
}
boost::shared_ptr<XsdTypeString> RoutingIdsAndExplicitDetails::getRoutingName()
{
   if(!this->routingNameIsNull_){
        if(routingNameIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeString>(IDManager::instance().getID(routingNameIDRef_));
        }else{
             return this->routingName_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeString>();
   }
}
boost::shared_ptr<Address> RoutingIdsAndExplicitDetails::getRoutingAddress()
{
   if(!this->routingAddressIsNull_){
        if(routingAddressIDRef_ != ""){
             return boost::shared_static_cast<Address>(IDManager::instance().getID(routingAddressIDRef_));
        }else{
             return this->routingAddress_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Address>();
   }
}
boost::shared_ptr<XsdTypeString> RoutingIdsAndExplicitDetails::getRoutingAccountNumber()
{
   if(!this->routingAccountNumberIsNull_){
        if(routingAccountNumberIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeString>(IDManager::instance().getID(routingAccountNumberIDRef_));
        }else{
             return this->routingAccountNumber_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeString>();
   }
}
std::vector<boost::shared_ptr<XsdTypeString>> RoutingIdsAndExplicitDetails::getRoutingReferenceText()
{
   if(!this->routingReferenceTextIsNull_){
        if(routingReferenceTextIDRef_ != ""){
             return this->routingReferenceText_;
        }else{
             return this->routingReferenceText_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<XsdTypeString>>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/RoutingIdsAndExplicitDetails.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class RoutingIdsAndExplicitDetails : ISerialized
    {
        public RoutingIdsAndExplicitDetails(XmlNode xmlNode)
        {
            XmlNodeList routingIdsNodeList = xmlNode.SelectNodes("routingIds");
            
            if (routingIdsNodeList != null)
            {
                this.routingIds_ = new List<RoutingIds>();
                foreach (XmlNode item in routingIdsNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            routingIdsIDRef_ = item.Attributes["id"].Value;
                            routingIds_.Add(new RoutingIds(item));
                            IDManager.SetID(routingIdsIDRef_, routingIds_[routingIds_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            routingIdsIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        routingIds_.Add(new RoutingIds(item));
                        }
                    }
                    else
                    {
                        routingIds_.Add(new RoutingIds(item));
                    }
                }
            }
            
        
            XmlNode routingNameNode = xmlNode.SelectSingleNode("routingName");
            
            if (routingNameNode != null)
            {
                if (routingNameNode.Attributes["href"] != null || routingNameNode.Attributes["id"] != null) 
                {
                    if (routingNameNode.Attributes["id"] != null) 
                    {
                        routingNameIDRef_ = routingNameNode.Attributes["id"].Value;
                        XsdTypeString ob = new XsdTypeString(routingNameNode);
                        IDManager.SetID(routingNameIDRef_, ob);
                    }
                    else if (routingNameNode.Attributes["href"] != null)
                    {
                        routingNameIDRef_ = routingNameNode.Attributes["href"].Value;
                    }
                    else
                    {
                        routingName_ = new XsdTypeString(routingNameNode);
                    }
                }
                else
                {
                    routingName_ = new XsdTypeString(routingNameNode);
                }
            }
            
        
            XmlNode routingAddressNode = xmlNode.SelectSingleNode("routingAddress");
            
            if (routingAddressNode != null)
            {
                if (routingAddressNode.Attributes["href"] != null || routingAddressNode.Attributes["id"] != null) 
                {
                    if (routingAddressNode.Attributes["id"] != null) 
                    {
                        routingAddressIDRef_ = routingAddressNode.Attributes["id"].Value;
                        Address ob = new Address(routingAddressNode);
                        IDManager.SetID(routingAddressIDRef_, ob);
                    }
                    else if (routingAddressNode.Attributes["href"] != null)
                    {
                        routingAddressIDRef_ = routingAddressNode.Attributes["href"].Value;
                    }
                    else
                    {
                        routingAddress_ = new Address(routingAddressNode);
                    }
                }
                else
                {
                    routingAddress_ = new Address(routingAddressNode);
                }
            }
            
        
            XmlNode routingAccountNumberNode = xmlNode.SelectSingleNode("routingAccountNumber");
            
            if (routingAccountNumberNode != null)
            {
                if (routingAccountNumberNode.Attributes["href"] != null || routingAccountNumberNode.Attributes["id"] != null) 
                {
                    if (routingAccountNumberNode.Attributes["id"] != null) 
                    {
                        routingAccountNumberIDRef_ = routingAccountNumberNode.Attributes["id"].Value;
                        XsdTypeString ob = new XsdTypeString(routingAccountNumberNode);
                        IDManager.SetID(routingAccountNumberIDRef_, ob);
                    }
                    else if (routingAccountNumberNode.Attributes["href"] != null)
                    {
                        routingAccountNumberIDRef_ = routingAccountNumberNode.Attributes["href"].Value;
                    }
                    else
                    {
                        routingAccountNumber_ = new XsdTypeString(routingAccountNumberNode);
                    }
                }
                else
                {
                    routingAccountNumber_ = new XsdTypeString(routingAccountNumberNode);
                }
            }
            
        
            XmlNodeList routingReferenceTextNodeList = xmlNode.SelectNodes("routingReferenceText");
            
            if (routingReferenceTextNodeList != null)
            {
                this.routingReferenceText_ = new List<XsdTypeString>();
                foreach (XmlNode item in routingReferenceTextNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            routingReferenceTextIDRef_ = item.Attributes["id"].Value;
                            routingReferenceText_.Add(new XsdTypeString(item));
                            IDManager.SetID(routingReferenceTextIDRef_, routingReferenceText_[routingReferenceText_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            routingReferenceTextIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        routingReferenceText_.Add(new XsdTypeString(item));
                        }
                    }
                    else
                    {
                        routingReferenceText_.Add(new XsdTypeString(item));
                    }
                }
            }
            
        
        }
        
    
        #region RoutingIds_
        private List<RoutingIds> routingIds_;
        public List<RoutingIds> RoutingIds_
        {
            get
            {
                if (this.routingIds_ != null)
                {
                    return this.routingIds_; 
                }
                else if (this.routingIdsIDRef_ != null)
                {
                    return this.routingIds_; 
                }
                else
                {
                      return this.routingIds_; 
                }
            }
            set
            {
                if (this.routingIds_ != value)
                {
                    this.routingIds_ = value;
                }
            }
        }
        #endregion
        
        public string routingIdsIDRef_ { get; set; }
        #region RoutingName_
        private XsdTypeString routingName_;
        public XsdTypeString RoutingName_
        {
            get
            {
                if (this.routingName_ != null)
                {
                    return this.routingName_; 
                }
                else if (this.routingNameIDRef_ != null)
                {
                    routingName_ = IDManager.getID(routingNameIDRef_) as XsdTypeString;
                    return this.routingName_; 
                }
                else
                {
                      return this.routingName_; 
                }
            }
            set
            {
                if (this.routingName_ != value)
                {
                    this.routingName_ = value;
                }
            }
        }
        #endregion
        
        public string routingNameIDRef_ { get; set; }
        #region RoutingAddress_
        private Address routingAddress_;
        public Address RoutingAddress_
        {
            get
            {
                if (this.routingAddress_ != null)
                {
                    return this.routingAddress_; 
                }
                else if (this.routingAddressIDRef_ != null)
                {
                    routingAddress_ = IDManager.getID(routingAddressIDRef_) as Address;
                    return this.routingAddress_; 
                }
                else
                {
                      return this.routingAddress_; 
                }
            }
            set
            {
                if (this.routingAddress_ != value)
                {
                    this.routingAddress_ = value;
                }
            }
        }
        #endregion
        
        public string routingAddressIDRef_ { get; set; }
        #region RoutingAccountNumber_
        private XsdTypeString routingAccountNumber_;
        public XsdTypeString RoutingAccountNumber_
        {
            get
            {
                if (this.routingAccountNumber_ != null)
                {
                    return this.routingAccountNumber_; 
                }
                else if (this.routingAccountNumberIDRef_ != null)
                {
                    routingAccountNumber_ = IDManager.getID(routingAccountNumberIDRef_) as XsdTypeString;
                    return this.routingAccountNumber_; 
                }
                else
                {
                      return this.routingAccountNumber_; 
                }
            }
            set
            {
                if (this.routingAccountNumber_ != value)
                {
                    this.routingAccountNumber_ = value;
                }
            }
        }
        #endregion
        
        public string routingAccountNumberIDRef_ { get; set; }
        #region RoutingReferenceText_
        private List<XsdTypeString> routingReferenceText_;
        public List<XsdTypeString> RoutingReferenceText_
        {
            get
            {
                if (this.routingReferenceText_ != null)
                {
                    return this.routingReferenceText_; 
                }
                else if (this.routingReferenceTextIDRef_ != null)
                {
                    return this.routingReferenceText_; 
                }
                else
                {
                      return this.routingReferenceText_; 
                }
            }
            set
            {
                if (this.routingReferenceText_ != value)
                {
                    this.routingReferenceText_ = value;
                }
            }
        }
        #endregion
        
        public string routingReferenceTextIDRef_ { get; set; }
        
    
        
    
    }
    
}

